File: employee-service-cpp/src/controllers/EmployeeController.cpp
#include "EmployeeController.h"
#include "Logger.h"
#include <nlohmann/json.hpp>

using json = nlohmann::json;

crow::response EmployeeController::getAllEmployees() {
    LOG_INFO("Request to get all employees");
    
    try {
        auto employees = EmployeeDatabase::getInstance().getAll();
        
        // Convert to JSON array
        json j = json::array();
        for (const auto& employee : employees) {
            j.push_back(employee.toJson());
        }
        
        LOG_INFO("Retrieved " + std::to_string(employees.size()) + " employees successfully");
        return crow::response(200, j.dump());
    } catch (const std::exception& e) {
        LOG_ERROR("Error retrieving all employees: " + std::string(e.what()));
        
        json error = {
            {"message", "Error retrieving employees"},
            {"error", e.what()}
        };
        
        return crow::response(500, error.dump());
    }
}

crow::response EmployeeController::getEmployeeById(const std::string& id) {
    LOG_INFO("Request to get employee with ID: " + id);
    
    try {
        auto employee = EmployeeDatabase::getInstance().getById(id);
        
        if (!employee) {
            LOG_WARNING("Employee with ID: " + id + " not found");
            
            json error = {
                {"message", "Employee with ID: " + id + " not found"}
            };
            
            return crow::response(404, error.dump());
        }
        
        LOG_INFO("Retrieved employee with ID: " + id + " successfully");
        return crow::response(200, employee->toJson().dump());
    } catch (const std::exception& e) {
        LOG_ERROR("Error retrieving employee with ID " + id + ": " + std::string(e.what()));
        
        json error = {
            {"message", "Error retrieving employee"},
            {"error", e.what()}
        };
        
        return crow::response(500, error.dump());
    }
}

crow::response EmployeeController::createEmployee(const crow::request& req) {
    LOG_INFO("Request to create a new employee");
    
    try {
        // Parse request body
        json body = json::parse(req.body);
        LOG_DEBUG("Request data: " + body.dump());
        
        // Validate input data
        Employee employee = Employee::fromJson(body);
        
        if (!Employee::isValidEmail(employee.getEmail())) {
            LOG_WARNING("Invalid email format: " + employee.getEmail());
            
            json error = {
                {"message", "Invalid employee data"},
                {"error", "Invalid email format"}
            };
            
            return crow::response(400, error.dump());
        }
        
        // Create employee
        auto newEmployee = EmployeeDatabase::getInstance().create(employee);
        
        LOG_INFO("Employee created successfully with ID: " + *newEmployee.getId());
        return crow::response(201, newEmployee.toJson().dump());
    } catch (const json::exception& e) {
        LOG_ERROR("Error parsing JSON: " + std::string(e.what()));
        
        json error = {
            {"message", "Invalid JSON data"},
            {"error", e.what()}
        };
        
        return crow::response(400, error.dump());
    } catch (const std::exception& e) {
        LOG_ERROR("Error creating employee: " + std::string(e.what()));
        
        json error = {
            {"message", "Error creating employee"},
            {"error", e.what()}
        };
        
        return crow::response(500, error.dump());
    }
}

crow::response EmployeeController::updateEmployee(const std::string& id, const crow::request& req) {
    LOG_INFO("Request to update employee with ID: " + id);
    
    try {
        // Parse request body
        json body = json::parse(req.body);
        LOG_DEBUG("Request data: " + body.dump());
        
        // Validate input data
        Employee employee = Employee::fromJson(body);
        
        if (!Employee::isValidEmail(employee.getEmail())) {
            LOG_WARNING("Invalid email format: " + employee.getEmail());
            
            json error = {
                {"message", "Invalid employee data"},
                {"error", "Invalid email format"}
            };
            
            return crow::response(400, error.dump());
        }
        
        // Update employee
        auto updatedEmployee = EmployeeDatabase::getInstance().update(id, employee);
        
        if (!updatedEmployee) {
            LOG_WARNING("Employee with ID: " + id + " not found");
            
            json error = {
                {"message", "Employee with ID: " + id + " not found"}
            };
            
            return crow::response(404, error.dump());
        }
        
        LOG_INFO("Employee with ID: " + id + " updated successfully");
        return crow::response(200, updatedEmployee->toJson().dump());
    } catch (const json::exception& e) {
        LOG_ERROR("Error parsing JSON: " + std::string(e.what()));
        
        json error = {
            {"message", "Invalid JSON data"},
            {"error", e.what()}
        };
        
        return crow::response(400, error.dump());
    } catch (const std::exception& e) {
        LOG_ERROR("Error updating employee with ID " + id + ": " + std::string(e.what()));
        
        json error = {
            {"message", "Error updating employee"},
            {"error", e.what()}
        };
        
        return crow::response(500, error.dump());
    }
}

crow::response EmployeeController::deleteEmployee(const std::string& id) {
    LOG_INFO("Request to delete employee with ID: " + id);
    
    try {
        bool result = EmployeeDatabase::getInstance().remove(id);
        
        if (!result) {
            LOG_WARNING("Employee with ID: " + id + " not found");
            
            json error = {
                {"message", "Employee with ID: " + id + " not found"}
            };
            
            return crow::response(404, error.dump());
        }
        
        LOG_INFO("Employee with ID: " + id + " deleted successfully");
        return crow::response(204);
    } catch (const std::exception& e) {
        LOG_ERROR("Error deleting employee with ID " + id + ": " + std::string(e.what()));
        
        json error = {
            {"message", "Error deleting employee"},
            {"error", e.what()}
        };
        
        return crow::response(500, error.dump());
    }
}

File: employee-service-python/app/controllers/employee_controller.py
from flask import jsonify, request
from http import HTTPStatus
from app.models.employee import EmployeeModel, EmployeeSchema
from loguru import logger

class EmployeeController:
    """Controller for Employee endpoints"""
    
    @staticmethod
    def get_all_employees():
        """Get all employees"""
        try:
            logger.info("Request to get all employees")
            employees = EmployeeModel.get_all()
            logger.info(f"Retrieved {len(employees)} employees successfully")
            return jsonify(employees), HTTPStatus.OK
        except Exception as e:
            logger.error(f"Error retrieving all employees: {str(e)}")
            return jsonify({"message": "Error retrieving employees", "error": str(e)}), HTTPStatus.INTERNAL_SERVER_ERROR
    
    @staticmethod
    def get_employee_by_id(employee_id):
        """Get employee by ID"""
        try:
            logger.info(f"Request to get employee with ID: {employee_id}")
            employee = EmployeeModel.get_by_id(employee_id)
            
            if not employee:
                logger.warning(f"Employee with ID: {employee_id} not found")
                return jsonify({"message": f"Employee with ID: {employee_id} not found"}), HTTPStatus.NOT_FOUND
            
            logger.info(f"Retrieved employee with ID: {employee_id} successfully")
            return jsonify(employee), HTTPStatus.OK
        except Exception as e:
            logger.error(f"Error retrieving employee with ID {employee_id}: {str(e)}")
            return jsonify({"message": "Error retrieving employee", "error": str(e)}), HTTPStatus.INTERNAL_SERVER_ERROR
    
    @staticmethod
    def create_employee():
        """Create a new employee"""
        try:
            logger.info("Request to create a new employee")
            
            # Get request data
            employee_data = request.json
            logger.debug(f"Request data: {employee_data}")
            
            # Validate input data
            try:
                EmployeeSchema(**employee_data)
            except Exception as e:
                logger.warning(f"Invalid employee data: {str(e)}")
                return jsonify({"message": "Invalid employee data", "error": str(e)}), HTTPStatus.BAD_REQUEST
            
            # Create employee
            new_employee = EmployeeModel.create(employee_data)
            logger.info(f"Employee created successfully with ID: {new_employee['id']}")
            
            return jsonify(new_employee), HTTPStatus.CREATED
        except Exception as e:
            logger.error(f"Error creating employee: {str(e)}")
            return jsonify({"message": "Error creating employee", "error": str(e)}), HTTPStatus.INTERNAL_SERVER_ERROR
    
    @staticmethod
    def update_employee(employee_id):
        """Update an employee"""
        try:
            logger.info(f"Request to update employee with ID: {employee_id}")
            
            # Get request data
            employee_data = request.json
            logger.debug(f"Request data: {employee_data}")
            
            # Update employee
            updated_employee = EmployeeModel.update(employee_id, employee_data)
            
            if not updated_employee:
                logger.warning(f"Employee with ID: {employee_id} not found")
                return jsonify({"message": f"Employee with ID: {employee_id} not found"}), HTTPStatus.NOT_FOUND
            
            logger.info(f"Employee with ID: {employee_id} updated successfully")
            return jsonify(updated_employee), HTTPStatus.OK
        except Exception as e:
            logger.error(f"Error updating employee with ID {employee_id}: {str(e)}")
            return jsonify({"message": "Error updating employee", "error": str(e)}), HTTPStatus.INTERNAL_SERVER_ERROR
    
    @staticmethod
    def delete_employee(employee_id):
        """Delete an employee"""
        try:
            logger.info(f"Request to delete employee with ID: {employee_id}")
            
            # Delete employee
            result = EmployeeModel.delete(employee_id)
            
            if not result:
                logger.warning(f"Employee with ID: {employee_id} not found")
                return jsonify({"message": f"Employee with ID: {employee_id} not found"}), HTTPStatus.NOT_FOUND
            
            logger.info(f"Employee with ID: {employee_id} deleted successfully")
            return "", HTTPStatus.NO_CONTENT
        except Exception as e:
            logger.error(f"Error deleting employee with ID {employee_id}: {str(e)}")
            return jsonify({"message": "Error deleting employee", "error": str(e)}), HTTPStatus.INTERNAL_SERVER_ERROR
